### PR TITLE
VMware: Allow user to specify VM version to latest

### LINKED ISSUE #56273: VMware: Feature request - vmware_guest Hardware Version - Latest
##### SUMMARY

We have a mixed environment made up of ESXi hosts ranging from version 6.0 to the latest version of 6.x, which means that the VM Hardware Version varies depending on the cluster we deploy to.

The Hardware: Version: only accepts VMX-XX where XX is a number between 3 and 15.  We need to ensure the Hardware version is the latest for the host the VM gets provisioned on.

Documentation:
https://pubs.vmware.com/vi3/sdk/ReferenceGuide/vim.VirtualMachine.html#upgradeVirtualHardware

##### ISSUE TYPE
- Feature Idea
What I'm proposing is the option to set Hardware: Version: to 'Latest' instead of specifying a number of VMX-3 through VMX-15  (The code i saw accepted 3 through 15 as acceptable).

##### COMPONENT NAME

vmware_guest

##### ADDITIONAL INFORMATION

In the existing code for vmware_guest, I noticed that after determining whether the number is within the 3-15 range,  at the very end it ultimately runs  vm_obj.UpgradeVM_Task(version) where version is an integer between 3 and 15.    What I'm proposing is that it be modified to potentially allow setting the value to 'Latest' instead of VMX-XX (where XX is a number between 3 and 15),  and as a result of specifying 'Latest' that the 'version' be set to Null,   so the resulting command would be    vm_obj.UpgradeVM_Task() which according to the vmware documentation, would upgrade the VM to the latest version supported by the host it's on.

```yaml

- name: Create a virtual machine on given ESXi hostname
  vmware_guest:
    hostname: "{{ vcenter_hostname }}"
    username: "{{ vcenter_username }}"
    password: "{{ vcenter_password }}"
    validate_certs: no
    folder: /DC1/vm/
    name: test_vm_0001
    state: poweredon
    guest_id: centos64Guest
    # This is hostname of particular ESXi server on which user wants VM to be deployed
    esxi_hostname: "{{ esxi_hostname }}"
    disk:
    - size_gb: 10
      type: thin
      datastore: datastore1
    hardware:
      memory_mb: 512
      num_cpus: 4
      scsi: paravirtual
      version: Latest  (instead of VMX-14, for example)
    networks:
    - name: VM Network
      mac: aa:bb:dd:aa:00:14
      ip: 10.10.10.100
      netmask: 255.255.255.0
      device_type: vmxnet3
    wait_for_ip_address: yes
  delegate_to: localhost
  register: deploy_vm

```

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
User now can specify VM Hardware version to latest.

Fixes: #56273

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/vmware/vmware_guest.py

##### ADDITIONAL INFORMATION
```yaml
    - name: deploy vm
      vmware_guest:
        hostname: "{{ vcenter_server }}"
        username: "{{ vcenter_user }}"
        password: "{{ vcenter_pass }}"
        validate_certs: no
        datacenter: "{{ datacenter }}"
        state: present
        folder: "/{{ datacenter }}/vm"
        name: VM_1448
        disk:
          - size: 10mb
            autoselect_datastore: yes
        hardware:
          memory_mb: 512
          num_cpus: 1
          version: latest
        guest_id: rhel7_64guest
```

### CODE REVIEW

**Comment by Akasurde** [comment]
cc @Tomorrow9

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 1]
```diff
@@ -1105,42 +1107,49 @@ def configure_hardware_params(self, vm_obj):
...
-                    if temp_version < version_digit:
-                        self.module.fail_json(msg="Current hardware version '%d' which is greater than the specified"
-                                                  " version '%d'. Downgrading hardware version is"
-                                                  " not supported. Please specify version greater"
-                                                  " than the current version." % (version_digit,
-                                                                                  temp_version))
-                    new_version = "vmx-%02d" % temp_version
-                    try:
-                        task = vm_obj.UpgradeVM_Task(new_version)
+                if temp_version == 'latest':
```
Comment (jillr) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py:1110]:
```suggestion
                if temp_version.lower() == 'latest':
```

**Review by jillr — CHANGES_REQUESTED** [review]
one suggestion, otherwise lgtm - working on 6.7.

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 2]
```diff
@@ -1105,42 +1107,49 @@ def configure_hardware_params(self, vm_obj):
...
+                    if vm_obj is not None:
```
Comment (Tomorrow9) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py:1111]: If 'template' parameter is set, and 'version' is also set, then 'vm_obj' is template, it's upgrade is not supported?

**Pull request closed without merging**